app/src/main/java/com/example/bottomnav/CAN_Data.java

```diff
@@ -0,0 +1,48 @@
+package com.example.bottomnav;
+
+class CAN_Data {
+    int id;
+    int[] data;
+
+    public CAN_Data(int id, int[] data, int len) {
+        this.id = id;
+        this.data = data;
+    }
+
+    public static CAN_Data decode(String raw) {
+        int id = -1;
+        int len;
+
+        int index = -1;
+        if (raw.charAt(0) == 't' || raw.charAt(0) == 'r') {
+            id = Integer.parseInt(raw.substring(1, 4), 16);
+            index = 4;
+        }
+        else if (raw.charAt(0) == 'T' || raw.charAt(0) == 'R') {
+            id = Integer.parseInt(raw.substring(1, 9), 16);
+            index = 9;
+        }
+        else {
+            return null;
+        }
+
+        len = Integer.parseInt(raw.substring(index, index + 1), 16);
```

Comment:
Consider adding a validation step to make sure:
- that the input is long enough such that it has a length field
- that the length is 8 or less
- that the input is long enough for the entire payload

and if it fails validation, return `Optional.empty()`

Also, here (and every other instance of `parseInt`) consider having error handling, specifically catching the `NumberFormatException` that `parseInt` can throw and returning Optional.empty(). You might as well wrap the entire method in a try-catch or something, since you use a lot of parseInt.